### PR TITLE
Change AssetScanner behaviour

### PR DESCRIPTION
**Pull request by mrtraser**
- add new callbacks
- now you can provide your own OCR request logic
- capture button customisation

### CODE REVIEW

**Comment by f1cognite** [comment]
This PR's storybook can now be previewed at the following URL: https://pr-263.gearboxjs.preview.cogniteapp.com

**Comment by f1cognite** [comment]
This PR's storybook can now be previewed at the following URL: https://pr-263.gearboxjs.preview.cogniteapp.com

File: src/components/AssetScanner/stories/full.md [hunk 1]
```diff
@@ -4,23 +4,23 @@
 
 #### Description:
 
-Using for recognising assets name strings from photo.
+Uses for recognising assets name strings from captured image.
```
Comment (xitman78) [inline src/components/AssetScanner/stories/full.md]:
```suggestion
The component can be used for recognizing the name of an asset in a captured image (from a webcam).
```

File: src/components/AssetScanner/stories/full.md [hunk 2]
```diff
@@ -4,23 +4,23 @@
...
 
 #### Usage requirements:
 
-Component depends on few third-part APIs, which is provided by Google
-Vision (to recognize text on picture). The user has to insert props
-to the component with a valid API key for Google Vision. Follow this
-guide to generate an api-key:
+By default, component uses Google Vision API (to recognize text on picture). User has to provide valid API key for Google Vision via **ocrKey** prop to use embedded OCR functional.
```
Comment (xitman78) [inline src/components/AssetScanner/stories/full.md]:
```suggestion
By default, the component uses Google Vision API (to recognize text on the picture). User has to provide a valid API key for Google Vision via **ocrKey** prop to use embedded OCR functional.
```

File: src/components/AssetScanner/stories/full.md [hunk 3]
```diff
@@ -4,23 +4,23 @@
...
 **Google Vision API**
 
-> Before you can use the Cloud Vision API, you must enable it for your project and generate your API key, see here how – [Enable the Vision API](https://cloud.google.com/vision/docs/before-you-begin).   
+> Before you can use the Cloud Vision API, you must enable it for your project and generate your API key, see here how – [Enable the Vision API](https://cloud.google.com/vision/docs/before-you-begin).
 
-> Don't forget for setting up API key restrictions – [Using API Keys](https://cloud.google.com/docs/authentication/api-keys).  
+> Don't forget for setting up API key restrictions – [Using API Keys](https://cloud.google.com/docs/authentication/api-keys).
 
-Google Vision API key could be pass to component via **ocrKey** prop or you can passURL address via **ocrUrl** prop in case of own backend realisation of Google Vision API call.  
+You might provide your custom `extractOcrStrings` function to retrieve string array from OCR recognition response structure.
```
Comment (xitman78) [inline src/components/AssetScanner/stories/full.md]:
```suggestion
You might provide your custom `extractOcrStrings` function to retrieve an array of strings from OCR recognition response structure.
```

File: src/components/AssetScanner/stories/full.md [hunk 4]
```diff
@@ -4,23 +4,23 @@
...
+By default, component uses Google Vision API (to recognize text on picture). User has to provide valid API key for Google Vision via **ocrKey** prop to use embedded OCR functional.
+Please, follow this guide to generate an api-key:
 
 **Google Vision API**
 
-> Before you can use the Cloud Vision API, you must enable it for your project and generate your API key, see here how – [Enable the Vision API](https://cloud.google.com/vision/docs/before-you-begin).   
+> Before you can use the Cloud Vision API, you must enable it for your project and generate your API key, see here how – [Enable the Vision API](https://cloud.google.com/vision/docs/before-you-begin).
 
-> Don't forget for setting up API key restrictions – [Using API Keys](https://cloud.google.com/docs/authentication/api-keys).  
+> Don't forget for setting up API key restrictions – [Using API Keys](https://cloud.google.com/docs/authentication/api-keys).
```
Comment (xitman78) [inline src/components/AssetScanner/stories/full.md]:
```suggestion
> Don't forget to set API key restrictions – [Using API Keys](https://cloud.google.com/docs/authentication/api-keys).
```

File: src/components/AssetScanner/stories/full.md [hunk 5]
```diff
@@ -4,23 +4,23 @@
...
-Vision (to recognize text on picture). The user has to insert props
-to the component with a valid API key for Google Vision. Follow this
-guide to generate an api-key:
+By default, component uses Google Vision API (to recognize text on picture). User has to provide valid API key for Google Vision via **ocrKey** prop to use embedded OCR functional.
+Please, follow this guide to generate an api-key:
 
 **Google Vision API**
 
-> Before you can use the Cloud Vision API, you must enable it for your project and generate your API key, see here how – [Enable the Vision API](https://cloud.google.com/vision/docs/before-you-begin).   
+> Before you can use the Cloud Vision API, you must enable it for your project and generate your API key, see here how – [Enable the Vision API](https://cloud.google.com/vision/docs/before-you-begin).
```
Comment (xitman78) [inline src/components/AssetScanner/stories/full.md]:
```suggestion
> Before you can use the Cloud Vision API, you must enable it in your project and generate your API key, see here how – [Enable the Vision API](https://cloud.google.com/vision/docs/before-you-begin).
```

File: src/components/AssetScanner/stories/full.md [hunk 6]
```diff
@@ -4,23 +4,23 @@
...
+Uses for recognising assets name strings from captured image.
 
 #### Usage requirements:
 
-Component depends on few third-part APIs, which is provided by Google
-Vision (to recognize text on picture). The user has to insert props
-to the component with a valid API key for Google Vision. Follow this
-guide to generate an api-key:
+By default, component uses Google Vision API (to recognize text on picture). User has to provide valid API key for Google Vision via **ocrKey** prop to use embedded OCR functional.
+Please, follow this guide to generate an api-key:
```
Comment (xitman78) [inline src/components/AssetScanner/stories/full.md]:
```suggestion
Please, follow this guide to generate an API-key:
```

**Comment by f1cognite** [comment]
This PR's storybook can now be previewed at the following URL: https://pr-263.gearboxjs.preview.cogniteapp.com

File: src/components/AssetScanner/stories/full.md [hunk 7]
```diff
@@ -4,23 +4,23 @@
...
+> Before you can use the Cloud Vision API, you must enable it for your project and generate your API key, see here how – [Enable the Vision API](https://cloud.google.com/vision/docs/before-you-begin).
 
-> Don't forget for setting up API key restrictions – [Using API Keys](https://cloud.google.com/docs/authentication/api-keys).  
+> Don't forget for setting up API key restrictions – [Using API Keys](https://cloud.google.com/docs/authentication/api-keys).
 
-Google Vision API key could be pass to component via **ocrKey** prop or you can passURL address via **ocrUrl** prop in case of own backend realisation of Google Vision API call.  
+You might provide your custom `extractOcrStrings` function to retrieve string array from OCR recognition response structure.
 
+You also can use your own implementation of OCR call if you use your own OCR service.
+Just provide `ocrRequest` function to get recognised from image strings with your own realisation of OCR request.
```
Comment (xitman78) [inline src/components/AssetScanner/stories/full.md]:
```suggestion
Just provide `ocrRequest` function to retrieve strings recognized from an image with your own implementation of OCR request.
```

File: src/components/AssetScanner/stories/full.md [hunk 8]
```diff
@@ -51,19 +46,34 @@ _No required props_
...
+| ------------------------ | ------------------------------------------------------------------------------------------------------------------------------------------------ | ---------------------------------------------- | ------------------ |
+| `button`                 | Render prop function, that should returns button node. Capture function and image base64 string are passed as arguments                          | `ButtonRenderProp`                             |                    |
+| `customNotification`     | Callback function to react on provided type of notification                                                                                      | `(type: ASNotifyTypes) => any`                 |                    |
+| `extractOcrStrings`      | Function, that get result of recognise function and format it as array of recognised strings (Can be used if you use embedded Google Vision API) | `(ocrResult: any) => string`                   |                    |
```
Comment (xitman78) [inline src/components/AssetScanner/stories/full.md]:
```suggestion
| `extractOcrStrings`      | Function that gets a result from the recognize function and formats the result as an array of recognized strings (Can be used if you use embedded Google Vision API) | `(ocrResult: any) => string`                   |                    |
```

File: src/components/AssetScanner/stories/full.md [hunk 9]
```diff
@@ -51,19 +46,34 @@ _No required props_
...
+| `ocrKey`                 | Api key property which is needed if you use embedded Google Vision API                                                                           | `string`                                       |                    |
```
Comment (xitman78) [inline src/components/AssetScanner/stories/full.md]:
```suggestion
| `ocrKey`                 | API key property which is needed if you use embedded Google Vision API                                                                           | `string`                                       |                    |
```

File: src/components/AssetScanner/stories/full.md [hunk 10]
```diff
@@ -51,19 +46,34 @@ _No required props_
...
+| `ocrRequest`             | Function that provide custom OCR call to detect strings on image                                                                                 | `(ocrParams: OcrRequest) => Promise<string[]>` |                    |
```
Comment (xitman78) [inline src/components/AssetScanner/stories/full.md]:
```suggestion
| `ocrRequest`             | Function that provides custom OCR call to detect strings on image                                                                                 | `(ocrParams: OcrRequest) => Promise<string[]>` |                    |
```

File: src/components/AssetScanner/stories/full.md [hunk 11]
```diff
@@ -51,19 +46,34 @@ _No required props_
...
+| `onStartLoading`         | Callback which triggers right after tacking shot from camera                                                                                     | `() => void`                                   |                    |
```
Comment (xitman78) [inline src/components/AssetScanner/stories/full.md]:
```suggestion
| `onStartLoading`         | Callback triggered right after taking a shot from camera                                                                                     | `() => void`                                   |                    |
```

File: src/components/AssetScanner/stories/full.md [hunk 12]
```diff
@@ -51,19 +46,34 @@ _No required props_
...
+| `onImageRecognizeStart`  | Callback which triggers when image recognition process start                                                                                     | `(image: string) => void`                      |                    |
```
Comment (xitman78) [inline src/components/AssetScanner/stories/full.md]:
```suggestion
| `onImageRecognizeStart`  | Callback triggered when image recognition process starts                                                                                     | `(image: string) => void`                      |                    |
```

File: src/components/AssetScanner/stories/full.md [hunk 13]
```diff
@@ -51,19 +46,34 @@ _No required props_
...
+| `onImageRecognizeFinish` | Callback which triggers when image recognition process finish and right before fetching assets by recognised strings                             | `(strings: string[] \| null) => void`          |                    |
```
Comment (xitman78) [inline src/components/AssetScanner/stories/full.md]:
```suggestion
| `onImageRecognizeFinish` | Callback triggered when image recognition process is finished and right before fetching assets by recognized strings                             | `(strings: string[] \| null) => void`          |                    |
```

File: src/components/AssetScanner/stories/full.md [hunk 14]
```diff
@@ -51,19 +46,34 @@ _No required props_
...
+| `onAssetFetchResult`     | Callback which triggers when sdk asset search finished                                                                                           | `(assets: Asset[]) => void`                    |                    |
```
Comment (xitman78) [inline src/components/AssetScanner/stories/full.md]:
```suggestion
| `onAssetFetchResult`     | Callback triggered when SDK asset search has been finished                                                                                           | `(assets: Asset[]) => void`                    |                    |
```

File: src/components/AssetScanner/stories/full.md [hunk 15]
```diff
@@ -51,19 +46,34 @@ _No required props_
...
+| `onEndLoading`           | Callback which triggers after finish of all recognition process                                                                                  | `() => void`                                   |                    |
```
Comment (xitman78) [inline src/components/AssetScanner/stories/full.md]:
```suggestion
| `onEndLoading`           | Callback triggered after finishing the recognition process                                                                                  | `() => void`                                   |                    |
```

File: src/components/AssetScanner/stories/full.md [hunk 16]
```diff
@@ -51,19 +46,34 @@ _No required props_
...
+| `onImageReset`           | Callback which triggers after resetting captured image                                                                                           | `() => void`                                   |                    |
```
Comment (xitman78) [inline src/components/AssetScanner/stories/full.md]:
```suggestion
| `onImageReset`           | Callback triggered after resetting captured image                                                                                           | `() => void`                                   |                    |
```

File: src/components/AssetScanner/stories/full.md [hunk 17]
```diff
@@ -51,19 +46,34 @@ _No required props_
...
+| `onOcrError`             | Callback which triggers when issues appears related to OCR service/request                                                                       | `(error: any) => void`                         |                    |
```
Comment (xitman78) [inline src/components/AssetScanner/stories/full.md]:
```suggestion
| `onOcrError`             | Callback triggered when occurs an error related to OCR service/request                                                                       | `(error: any) => void`                         |                    |
```

File: src/components/AssetScanner/stories/full.md [hunk 18]
```diff
@@ -51,19 +46,34 @@ _No required props_
...
+| `onError`                | Callback which triggers when some error occurs                                                                                                   | `(error: any) => void`                         |                    |
```
Comment (xitman78) [inline src/components/AssetScanner/stories/full.md]:
```suggestion
| `onError`                | Callback triggered when an error occurs                                                                                                   | `(error: any) => void`                         |                    |
```

File: src/utils/notifications.ts [hunk 19]
```diff
@@ -30,8 +30,8 @@ export const notification = ({
 
 export const ocrError = {
   type: NotificationTypes.ERROR,
-  message: 'Failed to scan image',
-  description: 'Check your API key configuration for Google Vision',
+  message: 'Some error appears',
```
Comment (xitman78) [inline src/utils/notifications.ts]:
```suggestion
  message: 'Network error',
```

File: src/utils/notifications.ts [hunk 20]
```diff
@@ -30,8 +30,8 @@ export const notification = ({
...
+  description: 'Check developer console for more info',
```
Comment (xitman78) [inline src/utils/notifications.ts]:
```suggestion
  description: 'Check developer\'s console for more info',
```

File: src/components/AssetScanner/stories/full.md (hunk 5 above)
Comment (mrtraser) [reply src/components/AssetScanner/stories/full.md]: "for" means that you have to enable Google Vision option in your Google Cloud account exactly FOR your project.

Comment (xitman78) [reply src/components/AssetScanner/stories/full.md]: got it

**Comment by f1cognite** [comment]
This PR's storybook can now be previewed at the following URL: https://pr-263.gearboxjs.preview.cogniteapp.com

**Comment by f1cognite** [comment]
This PR's storybook can now be previewed at the following URL: https://pr-263.gearboxjs.preview.cogniteapp.com

**Pull request merged**